unified-scheduler-logic/src/lib.rs

```diff
@@ -772,35 +772,22 @@ impl SchedulingStateMachine {
         index: usize,
         usage_queue_loader: &mut impl FnMut(Pubkey) -> UsageQueue,
     ) -> Task {
-        // Calling the _unchecked() version here is safe for faster operation, because
-        // `get_account_locks()` (the safe variant) is ensured to be called in
-        // DefaultTransactionHandler::handle() via Bank::prepare_unlocked_batch_from_single_tx().
-        //
-        // The safe variant has additional account-locking related verifications, which is crucial.
-        //
-        // Currently the replaying stage is redundantly calling `get_account_locks()` when unified
-        // scheduler is enabled on the given transaction at the blockstore. This will be relaxed
-        // for optimization in the future. As for banking stage with unified scheduler, it will
-        // need to run .get_account_locks() at least once somewhere in the code path. In the
-        // distant future, this function `create_task()` should be adjusted so that both stages do
-        // the checks before calling this (say, with some ad-hoc type like
-        // `SanitizedTransactionWithCheckedAccountLocks`) or do the checks here, resulting in
-        // eliminating the redundant one in the replaying stage and in the handler.
-        let locks = transaction.get_account_locks_unchecked();
-
-        let writable_locks = locks
-            .writable
+        // Locks are validated later in the pipeline. Here we create a task
+        // without checking that locks are valid.
```

Comment:
in fact, checking locks must be done prior to enqueuing tasks. that's because `SchedulingStateMachine` will cause dead locks on the existence of non-unique `UsageQeueue`s inside a given `Task`.

So, the original comment was wrong and misleading... Nevertheless, this comment has been serving as a reminder for me. how about retaining it while being updated like this?:

```rust
// It's crucial for tasks to be validated with
// `SanitizedTransaction::validate_account_locks()` prior to the creation.
// That's because it's part of protocol consensus regarding the rejection of
// blocks containing malformed transactions (`AccountLoadedTwice` and
// `TooManyAccountLocks`). Even more,`SchedulingStateMachine` can't properly
// handle transactions with duplicate addresses (those falling under
// `AccountLoadedTwice`).
//
// However, it's okay for now not to call `::validate_account_locks()` here. 
//
// Currently the replaying stage is always calling `get_account_locks()`
// regardless unified scheduler is enabled or not at the blockstore 
// (`Bank::prepare_sanitized_batch()` is called in `process_entries()`).
// This verification will be hoisted for optimization
// when removing `--block-verificatio-method=blockstore-processor`.
//
// As for banking stage with unified scheduler, it will need to run
// `.validate_account_locks()` at least once somewhere in the code path.
// In the distant future, this function (`create_task()`) should be adjusted
// so that both stages do the checks before calling this (say, with some ad-hoc
// type like `SanitizedTransactionWithCheckedAccountLocks`) or do the checks
// here, to simplify the two code paths regarding the essential
// `validate_account_locks` validation.
//
// Lastly, `validate_account_locks()` is currently called in
// `DefaultTransactionHandler::handle()` via
// `Bank::prepare_unlocked_batch_from_single_tx()` as well.
// This redundancy is known. It was just left as-is out of abundance of caution.
```

(please reword and/or reformat as needed...)

Reply:
108c35b5aada8409904d187a0aedd89a71e46f2c